README.md
parentrol
=========

Very simple parental control. Users have limited time on the computer.

Possibilities:
 * per-user control
 * start and and time, outside of which a user may not log in
 * maximum time per day
 * grace time (displays a popup warning the user that the max time or end time is almost reached)
 * screensaver aware
 * global, per-day and override controls + profiles
 * Gnome-shell extension to track the remaining time


debian/changelog
parentrol (0.0.6~rc5) unstable; urgency=low

  * fix /etc/default/parentrol
  * be sure that the parentroller.sh script stays running for each managed user
  * cleanup
  * logs cleanup
  * tighter perms
  * try to switch back to the active user if a user is killed while not active
  * Gnome3 shell extension to display the remaining time
  * temporary ban

 -- Cyril Adrian <cyril.adrian@gmail.com>  Fri, 10 Jun 2016 20:57:51 +0200

parentrol (0.0.5) unstable; urgency=low

  * Changes for Gnome 3.12
  * Added quit support to the user's parentroller
  * Moved logs to /var/log/parentrol.d

 -- Cyril Adrian <cyril.adrian@gmail.com>  Tue, 15 Jul 2014 15:50:46 +0200

parentrol (0.0.4) unstable; urgency=low

  * Better screensaver management (alas, relying on a user action)

 -- Cyril Adrian <cyril.adrian@gmail.com>  Sat, 04 Jan 2014 14:10:39 +0100

parentrol (0.0.3) unstable; urgency=low

  * Profiles management

 -- Cyril Adrian <cyril.adrian@gmail.com>  Mon, 21 Oct 2013 09:41:21 +0200

parentrol (0.0.2) unstable; urgency=low

  * Bash completion

 -- Cyril Adrian <cyril.adrian@gmail.com>  Sat, 28 Sep 2013 15:27:49 +0200

parentrol (0.0.1) unstable; urgency=low

  * Initial Release.

 -- Cyril Adrian <cyril.adrian@gmail.com>  Sun, 08 Sep 2013 20:21:33 +0200


gnome/extension.js
const Gio = imports.gi.Gio;
const GLib = imports.gi.GLib;
const Main = imports.ui.main;
const St = imports.gi.St;

let button, monitor;

function _refresh(sfile, text) {
   let result = function(monitor, a_file, other_file, event_type) {
      text.text = "...";
      if (event_type ==  Gio.FileMonitorEvent.DELETED) {
         text.text = "??:??";
      } else if (event_type == Gio.FileMonitorEvent.CHANGED || event_type == Gio.FileMonitorEvent.CREATED || event_type == Gio.FileMonitorEvent.MOVED) {
         let file = Gio.file_new_for_path(sfile);
         file.load_contents_async(null, function (source, result) {
            let [ok, contents, etag_out] = source.load_contents_finish(result);
            if (ok) {
               let left = parseInt(contents);
               let h = left / 60;
               let m = left % 60;
               if (h < 1 && m < 30) {
                  text.remove_style_class_name('counter-label-cool');
                  text.add_style_class_name('counter-label-hot');
               } else {
                  text.remove_style_class_name('counter-label-hot');
                  text.add_style_class_name('counter-label-cool');
               }
               text.text = "%02d:%02d".format(h, m);
            } else {
               text.text = "??:??";
            }
         });
      }
   };
   return result;
}

function init() {
   let sfile = "/tmp/parentroller/%s.left".format(GLib.get_user_name());
   button = new St.Bin({
      style_class: 'panel-button',
      reactive: false,
      can_focus: false,
      x_fill: true,
      y_fill: false,
      track_hover: false
   });
   let text = new St.Label({
      style_class: 'counter-label-cool',
      text: "Parentrol"
   });
   monitor = Gio.File.new_for_path(sfile).monitor_file(Gio.FileMonitorFlags.NONE, null);
   let refresh = _refresh(sfile, text);
   monitor.id = monitor.connect('changed', refresh);
   refresh(monitor, null, null, Gio.FileMonitorEvent.CREATED);
   button.set_child(text);
}

function enable() {
   Main.panel._rightBox.insert_child_at_index(button, 0);
}

function disable() {
   monitor.disconnect(monitor.id);
   Main.panel._rightBox.remove_child(button);
}


src/tools/check.sh
#!/bin/bash

# Parentrol: parental control
# Copyright (C) 2013-2016 Cyril Adrian <cyril.adrian@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


export ETCDIR=${ETCDIR:-/etc}
test -e $ETCDIR/default/parentrol && . $ETCDIR/default/parentrol

export DRY_RUN=${DRY_RUN:-true}
export ACTIVE=${ACTIVE:-true}
export TOOLSDIR=${TOOLSDIR:-$(dirname $(readlink -f $0))}
export LOCKDIR=${LOCKDIR:-$(dirname $(readlink -f $0))}

. $TOOLSDIR/_common.sh

$ACTIVE || {
    log "Not active"
    exit 0
}

test -d $ETCDIR/parentrol/users.d || {
    log "No users to watch"
    exit 0
}

/usr/bin/dotlockfile -l -r 3 -p $LOCKDIR/parentrol.lock || {
    log "**** parentrol locked! ($LOCKDIR/parentrol.lock)"
    exit 0
}

log "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
log "Starting parentrol checking session (NOW=$NOW DAY=$DAY)"

pids=""
for userdef in $(
    if [ -f $(readlink -f $ETCDIR/parentrol/profile) ]; then
        echo $ETCDIR/parentrol/profiles.d/$(< $ETCDIR/parentrol/profile)/users.d/*
    else
        echo $ETCDIR/parentrol/users.d/*
    fi
); do
    test -d $userdef && {
        user=$(basename $userdef)

        test -e /var/log/$user.log || {
            touch /var/log/$user.log
            chown $user:$user /var/log/$user.log
        }

        maxtime=$(cat_or_default $userdef/maxtime 120)
        gracetime=$(cat_or_default $userdef/gracetime 5)
        starttime=$(cat_or_default $userdef/starttime 0)
        endtime=$(cat_or_default $userdef/endtime $((24 * 60)))
        ban=$(cat_or_default $userdef/ban 0)

        check_user $user $maxtime $gracetime $starttime $endtime $ban &
        pids="$pids $!"
    }
done

wait $pids

/usr/bin/dotlockfile -u $LOCKDIR/parentrol.lock

log "Finished."


src/tools/_login_time.awk
#!/usr/bin/awk -f

# Parentrol: parental control
# Copyright (C) 2013-2016 Cyril Adrian <cyril.adrian@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


BEGIN {
    h = 24;
    m = 0;
    t = 0;
}

$2 ~ /(tty|:)[0-9]+/ && /still logged in/ {
    split($6, a, ":");
    H = strtonum(a[1]);
    M = strtonum(a[2]);
    if ((H < h) || (H == h && M < m)) {
        h = H;
        m = M;
    }
    next;
}

$2 ~ /(tty|:)[0-9]+/ && $7 == "-" && $8 ~ /[0-9]+:[0-9]+/ {
    split($6, a, ":");
    BH = strtonum(a[1]);
    BM = strtonum(a[2]);
    split($8, a, ":");
    EH = strtonum(a[1]);
    EM = strtonum(a[2]);
    t += (EH * 60 + EM) - (BH * 60 + BM);
    next;
}

END {
    if (h == 24) {
        t = strtonum(ENVIRON["NOW"]);
    } else {
        t += strtonum(ENVIRON["NOW"]) - (h * 60 + m);
    }
    printf("%d\n", t);
}


src/tools/parentroller.sh
#!/bin/bash

# Parentrol: parental control
# Copyright (C) 2013-2016 Cyril Adrian <cyril.adrian@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# NOTE:
# This script must be started by each user that needs to be controlled.

if [ $(id -u) == 0 ]; then
    echo "Cannot run as root!" >&2
    exit 1
fi

export ETCDIR=${ETCDIR:-/etc}
test -e $ETCDIR/default/parentrol && . $ETCDIR/default/parentrol
export TOOLSDIR=${TOOLSDIR:-$(dirname $(readlink -f $0))}
export PARENTROLLER_DIR=${PARENTROLLER_DIR:-/tmp/parentroller}
export PARENTROLLER_LOGDIR=${PARENTROLLER_LOGDIR:-/var/log/parentrol.d}

function check_screensaver {
    #gnome-screensaver-command -q
    dbus-send --session --dest=org.gnome.ScreenSaver --type=method_call --print-reply --reply-timeout=5000 /org/gnome/ScreenSaver org.gnome.ScreenSaver.GetActive
}

case x"$1" in
    x-saver)
        if [ ! -r "$PARENTROLLER_DIR/$2" ]; then
            echo "File $PARENTROLLER_DIR/$2 not found!" >&2
            exit 1
        fi

        file=$(<"$PARENTROLLER_DIR/$2")
        if [ -e "$file" ]; then
            echo "File $file must not exist!" >&2
            exit 1
        fi
        check_screensaver > "$file" 2>&1
        rm -f "$PARENTROLLER_DIR/$2".lock
        ;;

    x-quit)
        if [ ! -r "$PARENTROLLER_DIR/$2" ]; then
            echo "File $PARENTROLLER_DIR/$2 not found!" >&2
            exit 1
        fi
        echo $(date -R)": Logging out of GNOME session"
        killall yad
        gnome-session-quit --logout --force --no-prompt
        ;;

    x)
        user=$(id -un)
        log="$PARENTROLLER_LOGDIR/$user.log"
        test -w $log || {
            echo "Cannot log to $log" >&2
            exit 1
        }

        {
            echo
            echo $(date -R)": Starting parentroller"
            rm -f "$PARENTROLLER_DIR/$user.saver" "$PARENTROLLER_DIR/$user.quit"
            inoticoming --logfile "$log" "$PARENTROLLER_DIR" --stdout-to-log --stderr-to-log --regexp "^$user.saver$" $0 -saver {} \;
            inoticoming --logfile "$log" "$PARENTROLLER_DIR" --stdout-to-log --stderr-to-log --regexp "^$user.quit$" $0 -quit {} \;

            ext_uuid='ParentrolView@cadrian.net'
            ext_list=$(gsettings get org.gnome.shell enabled-extensions | sed 's/^\[\(.*\)\]$/\1/')
            echo $ext_list | grep -q "$ext_uuid" || {
                gsettings set org.gnome.shell enabled-extensions "[${ext_list}, '${ext_uuid}']"
                killall -HUP gnome-shell #gnome-shell --replace & disown
            }
        } >>$log 2>&1

        # The checker expects parentroller to be running.
        # Make the process wait without using CPU:
        run="$PARENTROLLER_DIR/$user.run"
        rm -f "$run" && mkfifo "$run"
        exec 3<> "$run" # fifo opened RW but never written to by this process; hence, wait...
        while true; do
            read r <&3
            echo "|$r" >>$log
        done
        ;;

    *)
        echo "Unknown argument: $1" >&2
        exit 1
        ;;
esac


src/tools/_common.sh
# to be sourced

# Parentrol: parental control
# Copyright (C) 2013-2016 Cyril Adrian <cyril.adrian@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


umask 077
export LANG=C
export NOW=$(date +'%H * 60 + %M' | bc)
export DAY=$(date +'%a')
export LOG=${LOG:-/var/log/parentrol.d/check.log}
export TMPDIR=${TMPDIR:-/tmp}/parentrol.$(id -u)
export PARENTROLLER_DIR=${PARENTROLLER_DIR:-/tmp/parentroller}
export PARENTROLLER_LOGDIR=${PARENTROLLER_LOGDIR:-/tmp/parentroller}

mkdir -p $TMPDIR $PARENTROLLER_DIR $(dirname $LOG)
chmod 1777 $PARENTROLLER_DIR
chmod 755 $(dirname $LOG)
chmod 644 $LOG

function log {
    echo $(date +'%Y/%m/%d %H:%M:%S') "$@" >> $LOG
}

function check_parentroller {
    local user=$1
    local ck_nb=2
    local ret=1

    login_date=$(last -R $user | grep "still logged in" | awk '$2 ~ /:[0-9]+/ {split($6, a, ":"); print strtonum(a[1]) * 60 + strtonum(a[2])}')

    while [ $ck_nb -gt 0 ]; do
        log "Check parentroller for $user - #$ck_nb"
        if ps axu | egrep "^$user[[:space:]]" | cut -c66- | egrep -q '^(/bin/bash )?'${TOOLSDIR%/}'/parentroller.sh$'; then
            ret=0
            ck_nb=0
        elif [ $NOW -gt $(($login_date + 1)) ]; then
            echo "Parentroller for user $user seems not to be running (in $TOOLSDIR)!" >&2 # will be mailed by cron
            log "Error: parentroller for user $user seems not to be running."
            ck_nb=0
        else
            # $NOW is only slightly greater than $login_date, maybe the user just logged in and the parentroller is not yet started.
            log "Warning: parentroller for user $user seems not to be running (just logged in?) Waiting a bit."
            ck_nb=$(($ck_nb - 1))
            if [ $ck_nb -gt 0 ]; then
                sleep 30
            fi
        fi
    done
    return $ret
}

function ensure_parentroller {
    local user=$1
    local desktop=/home/$user/.config/autostart/parentroller.sh.desktop

    test -e $desktop || {
        cat > $desktop <<EOF
[Desktop Entry]
Type=Application
Name=Parentroller
Exec=${TOOLSDIR%/}/parentroller.sh
Hidden=false
NoDisplay=false
Terminal=false
StartupNotify=false
X-GNOME-Autostart-enabled=true
Comment=Parentroller
EOF
    }
    chmod 750 $desktop
    chown $user:$user $desktop

    log="$PARENTROLLER_LOGDIR/$user.log"
    touch "$log"
    chown $user:$user "$log"
    chmod 644 "$log"
}

function get_user_display {
    local user=$1
    local tty

    tty=$(last -R $user | grep "still logged in" | awk '$2 ~ /:[0-9]+/ {print $2}') && {
        test -n "$tty" && {
            echo $tty
            return 0
        }
    }
    tty=$(last -R $user | grep "still logged in" | awk '$2 ~ /tty[0-9]+/ {print $2}') && {
        test -n "$tty" && {
            ps -f -C Xorg | awk '$6 == "'$tty'" { print $9 }'
            return 0
        }
    }
    echo "not found"
    return 1
}

function kill_user_now {
    local user=$1
    local display

    shift
    log "**** Kill user: $user ($@)"
    display=$(get_user_display $user) && {
        $DRY_RUN || {
            passwd -lq $user
            {
                check_parentroller $user && {
                    test -p $PARENTROLLER_DIR/$user.run && echo "Parentrol: ask quit $user" >> $PARENTROLLER_DIR/$user.run
                    touch $PARENTROLLER_DIR/$user.quit
                    sleep 10
                }
                slay -clean $user
            } >/dev/null 2>&1
            echo "User $user slain! ($@)" >&2 # will be mailed by cron
        }
    }
}

function kill_user {
    local user=$1
    local display

    shift
    display=$(get_user_display $user) && {
        $DRY_RUN || {
            local active_vt=$(fgconsole)
            local user_vt=$(grep "using VT number" /var/log/Xorg.${display#:}.log | egrep -o '[0-9]+$')
            log "Must kill $user ($@) -- active VT is $active_vt / user VT is $user_vt"
            if [ $active_vt -ne $user_vt ]; then
                if [ -e $TMPDIR/$user.slay ]; then
                    log "User $user already warned"
                else
                    echo "User $user should be killed; waiting because s/he is not active."
                    touch $TMPDIR/$user.slay
                    su $user -c "DISPLAY=$display yad --title 'ATTENTION' --text='SESSION EXPIRÉE' --no-buttons --undecorated --fullscreen --sticky --center --on-top --justify=center" &
                    disown
                fi
            else
                kill_user_now $user "$@"
            fi
        }
    }
}

function warn_user {
    local user=$1
    local gracetime=$2
    local display

    log "**** Warn user: $user"
    display=$(get_user_display $user) && {
        $DRY_RUN || {
            su $user -c "DISPLAY=$display yad --title 'ATTENTION' --text='FIN DE SESSION DANS $gracetime MINUTES' --button=gtk-ok:0 --undecorated --sticky --center --on-top --justify=center" &
            disown
        }
    }
}

function check_screensaver {
    local user=$1
    local display
    local saver
    local lock

    log "Checking screensaver of $user"
    display=$(get_user_display $user) && {
        check_parentroller $user || {
            log "Considering screensaver of $user inactive."
            return 1
        }

        local active_vt=$(fgconsole)
        local user_vt=$(grep "using VT number" /var/log/Xorg.${display#:}.log | egrep -o '[0-9]+$')
        log "$user's VT is $user_vt"
        if [ $active_vt -ne $user_vt ]; then
            log "Active VT is $active_vt, obviously $user is not active."
            # Hence behave as if the screensaver were active
            return 0
        fi

        saver=$PARENTROLLER_DIR/$user.saver
        rm -f $saver*
        lock=$saver.lock
        /usr/bin/dotlockfile -l -r 3 $lock || {
            log "Error: $lock already existing?? Ignoring"
        }
        chown $user:$user $lock # will be removed by parentroller, therefore it must belong to the right user
        echo $saver.data > $saver # triggers inotify => the user's parentroller
        chmod a+r $saver

        /usr/bin/dotlockfile -l -r 3 $lock || {
            log "$lock not removed by parentroller of $user... taking too long? Considering screensaver inactive."
            rm -f $saver*
            return 1
        }
        log $(ls $saver*)
        test -r $saver.data || {
            log "No data returned by parentroller of $user. Considering screensaver inactive."
            rm -f $saver*
            return 1
        }
        while read line; do
            log " | $line"
        done < $saver.data
        if grep -q "boolean true" $saver.data || grep -q "The screensaver is active" $saver.data; then
            log "Screensaver is running for $user!"
            rm -f $saver*
            return 0
        elif grep -q '^Error org.freedesktop.DBus.Error.NoReply:' $saver.data; then
            # Looks like the dbus daemon refuses to answer when not on the right console.
            # If we are not on the right console, we consider that the use is not active.
            log "Screensaver check failed for $user."
            rm -rf $saver*
            return 0
        fi
        rm -f $saver*
    }
    return 1
}

function count_screensaver {
    local user=$1
    local file
    local ss_count

    file=$TMPDIR/$user.screensaver
    if [ -e $file ]; then
        ss_count=$(<$file)
    else
        ss_count=0
    fi

    if check_screensaver $user ; then
        ss_count=$(($ss_count + 1))
        log "screensaver is active for $user: $ss_count"
        echo $ss_count > $file
        return 1
    fi

    log "screensaver count for $user: $ss_count"
    echo $ss_count
    return 0
}

function check_logged_in_user {
    local user=$1
    local maxtime=$2
    local gracetime=$3
    local starttime=$4
    local endtime=$5
    local ban=$6
    local ss_count
    local login_time
    local ret=0

    login_time=$(
        last -R $user | \
            grep "$(date +'%a %b %_d')" | \
            while read l; do
                log " + $l"
                echo "$l"
            done | \
            $TOOLSDIR/_login_time.awk
    )

    log "$user: login_time=$login_time"

    if [ $NOW -lt $starttime ]; then
        kill_user $user "too early"
    elif [ $NOW -gt $endtime ]; then
        kill_user $user "too late"
    elif [ $ban -gt $(date +'%s') ]; then
        kill_user_now $user "user is banned until $(date -R -d"@$ban")"
    elif ss_count=$(count_screensaver $user); then
        time_left=$(
            t1=$(($maxtime + $ss_count - $login_time))
            t2=$(($endtime - $NOW))
            if [ $t1 -lt $t2 ]; then
                echo $t1
            else
                echo $t2
            fi
        )
        su $user -c "echo $time_left > $PARENTROLLER_DIR/$user.left"

        if [ $time_left -lt 0 ]; then
            kill_user $user "time expired"
        elif [ $time_left -lt $gracetime ]; then
            log "$user has $time_left minutes left!"
            if [ -e $TMPDIR/$user.warn ]; then
                log "$user already warned"
            else
                touch $TMPDIR/$user.warn
                warn_user $user $gracetime
            fi
        else
            log "$user has $time_left minutes left"
            ret=1
        fi
    fi

    return $ret
}

function check_user {
    local user=$1
    local maxtime=$2
    local gracetime=$3
    local starttime=$4
    local endtime=$5
    local ban=$6

    log "Checking $user ($starttime-$endtime: max $maxtime/$gracetime)"
    log "Display of $user is" $(get_user_display $user)
    last -R $user | grep "still logged in" | while read line; do
        log " | $line"
    done

    ensure_parentroller $user

    if last -R $user | grep -v "$(date +'%a %b %_d')" | grep -q "still logged in" ; then
        kill_user_now $user "still logged in since yesterday"
    elif last -R $user | grep -q "still logged in" ; then
        if check_logged_in_user $user $maxtime $gracetime $starttime $endtime $ban ; then
            return 0
        fi
    fi

    if last -R $user | grep -q "$(date +'%a %b %_d')" ; then
        log "$user logged in and out today"
    elif [ $NOW -lt $starttime ]; then
        log "$user cannot log in yet (too early)"
        $DRY_RUN || passwd -lq $user
    elif [ $NOW -gt $endtime ]; then
        log "$user cannot log in anymore (too late)"
        $DRY_RUN || passwd -lq $user
    else
        log "$user allowed, not logged in yet today"

        # (always do it to return to sane defaults)
        $DRY_RUN || passwd -uq $user
        rm -f $TMPDIR/$user.warn $TMPDIR/$user.slay $TMPDIR/$user.screensaver $PARENTROLLER_DIR/$user.left
    fi
}

function cat_or_default {
    local file=$1
    local default=$2

    if [ -e $file.ovr ]; then
        cat $file.ovr
    elif [ -e $file.$DAY ]; then
        cat $file.$DAY
    elif [ -e $file ]; then
        cat $file
    else
        echo $default
    fi
}
